src/System.Linq.Parallel/tests/Combinatorial/ParallelQueryCombinationTests.cs

```diff
@@ -0,0 +1,1523 @@
+            Assert.Equal(defaultStart + defaultSize, seen);
+        }
+
+        [Theory]
+        [MemberData("UnaryOperators")]
+        [MemberData("BinaryOperators")]
+        [MemberData("UnaryUnorderedOperators")]
+        [MemberData("BinaryUnorderedOperators")]
+        public static void ThenByDescending_OtherDirection(LabelledOperation source, LabelledOperation operation)
+        {
+            int seen = defaultStart + defaultSize;
+            foreach (int i in operation.Item(defaultStart, defaultSize, source.Item).OrderBy(x => 0).ThenByDescending(x => x))
+            {
+                Assert.Equal(--seen, i);
+            }
+            Assert.Equal(defaultStart, seen);
+        }
+
+        [Theory]
+        [MemberData("UnaryOperators")]
+        [MemberData("BinaryOperators")]
+        [MemberData("UnaryUnorderedOperators")]
+        [MemberData("BinaryUnorderedOperators")]
+        public static void ThenByDescending_OtherDirection_NotPipelined(LabelledOperation source, LabelledOperation operation)
+        {
+            int seen = defaultStart + defaultSize;
+            Assert.All(operation.Item(defaultStart, defaultSize, source.Item).OrderBy(x => 0).ThenByDescending(x => x).ToList(), x => Assert.Equal(--seen, x));
+            Assert.Equal(defaultStart, seen);
+        }
+
+        [Theory]
+        [MemberData("UnaryOperators")]
+        [MemberData("BinaryOperators")]
+        public static void ToArray(LabelledOperation source, LabelledOperation operation)
+        {
+            int seen = defaultStart;
+            Assert.All(operation.Item(defaultStart, defaultSize, source.Item).ToArray(), x => Assert.Equal(seen++, x));
+            Assert.Equal(defaultStart + defaultSize, seen);
+        }
+
+        [Theory]
+        [MemberData("UnaryFailingOperators")]
+        [MemberData("BinaryFailingOperators")]
+        [MemberData("OrderFailingOperators")]
+        public static void ToArray_AggregateException(LabelledOperation source, LabelledOperation operation)
+        {
+            AggregateException ae = Assert.Throws<AggregateException>(() => operation.Item(defaultStart, defaultSize, source.Item).ToArray());
+            Assert.All(ae.InnerExceptions, e => Assert.IsType<DeliberateTestException>(e));
+        }
+
+        [Theory]
+        [MemberData("UnaryOperators")]
+        [MemberData("BinaryOperators")]
+        [MemberData("UnaryUnorderedOperators")]
+        [MemberData("BinaryUnorderedOperators")]
+        public static void ToDictionary(LabelledOperation source, LabelledOperation operation)
+        {
+            IntegerRangeSet seen = new IntegerRangeSet(defaultStart, defaultSize);
+            Assert.All(operation.Item(defaultStart, defaultSize, source.Item).ToDictionary(x => x * 2),
+                p =>
+                {
+                    Assert.True(seen.Add(p.Key / 2));
+                    Assert.Equal(p.Key, p.Value * 2);
+                });
+            Assert.True(seen.Complete());
```

Comment:
In all of the places you do ```Assert.True(seen.Complete())```, did you consider instead just having a call like ```seen.AssertComplete()```, where the implementation would include in the assert message information about what was missing (or even duplicated if you chose to change the implementation of the collection)?